kernel/src/snake.rs

```diff
@@ -0,0 +1,366 @@
+        self.grid.set(generate_apple_pos(&self.grid), Cell::Apple);
+    }
+
+    fn notification(&mut self, message: &str) {
+        let old_color = STDOUT.read().color();
+        STDOUT.write().set_color(color!(White on Black)).expect("Error setting color!");
+        let center = STDOUT.read().resolution().center();
+
+        centered_text(message, center.x, center.y);
+        centered_text("Press any key to continue...", center.x, center.y - 1);
+
+        STDOUT.write().set_color(old_color).expect("Error setting color!");
+
+        pit::sleep(1000);
+
+        loop {
+            if let Ok(Some(event)) = self.keyboard.read_event() {
+                if event.event_type == KeyEventType::Break {
+                    break;
+                }
+            }
+        }
+    }
+}
+
+struct Grid {
+    cells: Vec<Cell>,
+    width: usize,
+    height: usize,
+}
+
+impl Grid {
+    fn empty(width: usize, height: usize) -> Grid {
+        Grid {
+            cells: vec![Cell::Empty; width * height],
+            width,
+            height,
+        }
+    }
+
+    fn clear(&mut self) {
+        for cell in self.cells.iter_mut() {
+            *cell = Cell::Empty;
+        }
+    }
+
+    fn set(&mut self, point: Point, cell: Cell) {
+        self.cells[Grid::index(point)] = cell;
+        STDOUT.write().set_char(cell.character(), point)
+            .expect("failed to draw cell to screen");
+    }
+
+    fn get(&self, point: Point) -> &Cell {
+        &self.cells[Grid::index(point)]
+    }
+
+    #[inline]
+    fn index(point: Point) -> usize {
+        if point.x >= 80 || point.y >= 25 {
+            panic!("point out of bounds {} {}", point.x, point.y);
+        }
+        point.x + point.y * 80
+    }
+}
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+enum Cell {
+    Empty,
+    Head,
+    Body,
+    Apple,
+}
+
+impl Cell {
+    fn character(&self) -> TerminalCharacter {
+        match self {
+            Cell::Empty => TerminalCharacter::new(' ', color!(Black on Black)),
+            Cell::Head => TerminalCharacter::new(HEAD_CHAR, color!(White on Black)),
+            Cell::Body => TerminalCharacter::new(' ', color!(Green on Green)),
+            Cell::Apple => TerminalCharacter::new(' ', color!(Red on Red))
+        }
+    }
+}
+
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+enum Direction {
+    Up,
+    Down,
+    Left,
+    Right,
+}
+
+struct Snake {
+    head: Point,
+    direction: Direction,
+    blocks: Vec<Point>,
+    len: u16,
+    first_tick: bool,
+}
+
+impl Snake {
+    fn new() -> Snake {
+        Snake {
+            head: STDOUT.read().resolution().center(),
+            direction: Direction::Right,
+            blocks: Vec::with_capacity(128),
+            len: 4,
+            first_tick: true,
+        }
+    }
+
+    fn update(&mut self, grid: &mut Grid) -> MoveResult {
+        // Replace the previous head position with a snake body block
+        if self.first_tick {
```

Comment:
I feel like `first_tick` is a bit of a hack solution -- maybe instead check that `blocks.len() >0`?

Reply:
Does it matter?